src/deploy/D3MInit.sol

```diff
@@ -0,0 +1,266 @@
+}
+
+interface CDaiLike {
+    function interestRateModel() external view returns (address);
+    function implementation() external view returns (address);
+}
+
+interface D3MOracleLike {
+    function vat() external view returns (address);
+    function ilk() external view returns (bytes32);
+    function file(bytes32, address) external;
+}
+
+interface D3MHubLike {
+    function vat() external view returns (address);
+    function daiJoin() external view returns (address);
+    function file(bytes32, address) external;
+    function file(bytes32, bytes32, address) external;
+    function file(bytes32, bytes32, uint256) external;
+}
+
+interface D3MMomLike {
+    function setAuthority(address) external;
+}
+
+struct D3MCommonConfig {
+    bytes32 ilk;
+    bool existingIlk;
+    uint256 maxLine;
+    uint256 gap;
+    uint256 ttl;
+    uint256 tau;
+}
+
+struct D3MAaveConfig {
+    address king;
+    uint256 bar;
+    address adai;
+    address stableDebt;
+    address variableDebt;
+    address tack;
+    uint256 adaiRevision;
+}
+
+struct D3MCompoundConfig {
+    address king;
+    uint256 barb;
+    address cdai;
+    address comptroller;
+    address comp;
+    address tack;
+    address delegate;
+}
+
+// Init a D3M instance
+library D3MInit {
+
+    function initCore(
+        DssInstance memory dss,
+        D3MCoreInstance memory d3mCore
+    ) internal {
+        D3MHubLike hub = D3MHubLike(d3mCore.hub);
+        D3MMomLike mom = D3MMomLike(d3mCore.mom);
+
+        // Sanity checks
+        require(hub.vat() == address(dss.vat), "Hub vat mismatch");
+        require(hub.daiJoin() == address(dss.daiJoin), "Hub daiJoin mismatch");
+
+        hub.file("vow", address(dss.vow));
+        hub.file("end", address(dss.end));
+
+        mom.setAuthority(dss.chainlog.getAddress("MCD_ADM"));
+
+        dss.vat.rely(address(hub));
+
+        dss.chainlog.setAddress("DIRECT_HUB", address(hub));
+        dss.chainlog.setAddress("DIRECT_MOM", address(mom));
+    }
+
+    function _init(
+        DssInstance memory dss,
+        D3MInstance memory d3m,
+        D3MCommonConfig memory cfg,
+        address gem
+    ) private {
+        bytes32 ilk = cfg.ilk;
+        D3MHubLike hub = D3MHubLike(dss.chainlog.getAddress("DIRECT_HUB"));
+        D3MOracleLike oracle = D3MOracleLike(d3m.oracle);
+
+        // Sanity checks
+        require(oracle.vat() == address(dss.vat), "Oracle vat mismatch");
+        require(oracle.ilk() == ilk, "Oracle ilk mismatch");
+
+        hub.file(ilk, "pool", d3m.pool);
+        hub.file(ilk, "plan", d3m.plan);
+        hub.file(ilk, "tau", cfg.tau);
+
+        oracle.file("hub", address(hub));
+
+        dss.spotter.file(ilk, "pip", address(oracle));
+        dss.spotter.file(ilk, "mat", 10 ** 27);
+        if (!cfg.existingIlk) {
+            dss.vat.init(ilk);
+            dss.jug.init(ilk);
+        }
+        dss.vat.file(ilk, "line", cfg.gap);
+        dss.vat.file("Line", dss.vat.Line() + cfg.gap);
```

Comment:
This should take into account if there was already a ilk.line before and discount it.

Reply:
done